Gh-534: Initial docs for federated POC (#535)
gchq/gaffer-doc: docs/administration-guide/gaffer-stores/simple-federated/additional-info.md

# Additional Information on Simple Federation

This page contains additional information and considerations

> a admin may need to know when using the federated store type.

Reviewer: ```suggestion
an admin may need to know when using the federated store type.
```

| `federated.excludedGraphIds` | List of graph IDs to exclude from the query. If this is set any graph IDs on a `federated.graphIds` option are ignored and instead, all graphs are executed on except the ones specified e.g. `"graph1,graph2"` |
| `federated.aggregateElements` | Should the element aggregator be used when merging element results. |
| `federated.forwardChain` | Should the whole operation chain be sent to the sub graph or not. If set to `false` each operation will inside the chain will be sent separately, so merging from each graph will happen after each operation instead of at the end of the chain. This will be inherently slower if turned off so is `true` by default. |

Along with the options above, all merge classes can be overridden per query
using the same property key as you would via the store properties. Please see
the table [here](./configuration.md#store-properties) for more information.

If you wish to submit different operations to different graphs in the same query
you can do this using the `federate.forwardChain` option. By setting this to
false on the outer operation chain the options on the operations inside it will

> be honoured. An example of this can been below:

Reviewer: ```suggestion
be honoured. An example of this can be seen below:
```

                "class": "GetElements",
                "options": {
                    "federated.graphIds": "graph1"
                },
                "input": [
                    {
                        "class": "EntitySeed",
                        "vertex": "1"
                    }
                ]
            },
            {
                "class": "AddElements",
                "options": {
                    "federated.graphIds": "graph2"
                }
            }
        ]
    }
    ```

## Cache Considerations

The federated store utilises the [Gaffer cache](../store-guide.md#caches) to store
graphs that have been added to the store. This means all features available to
normal caches are also available to the graph storage, allowing the sharing and
persisting of graphs between instances.

The federated store will use the default cache service to store graphs in. It
will also add a standard suffix meaning if you want to share graphs you will
need to set this to something other than the graph ID (see [here](../store-guide.md#cache-service)).

## Schema Compatibility

When querying multiple graphs the schemas of each graph the federated store will

> attempt to merge them together. This means the schemas will need to be

Reviewer: ? something like
```suggestion
When querying multiple graphs, the federated store will attempt to merge each graph's schema together. This means the schemas will need to be
```

compatible in order to query across them. Generally you will need to ensure

> any shared groups can be merged correctly a few examples of criteria to

Reviewer 2: ```suggestion
any shared groups can be merged correctly, a few examples of criteria to
```

consider are:

- Any properties in a shared group defined in both schemas need to have the same
  type and aggregation function.
- Any visibility properties need to be compatible or they will be removed from the
  schema.

> - Group with different properties in each schema will be merged so the group has

Reviewer: ```suggestion
- Groups with different properties in each schema will be merged so the group has
```

  all the properties in the merged schema.

> - Any group by definitions need to be compatible or will be removed.

Reviewer: ? assuming this is what you meant
```suggestion
- Any groupBy definitions need to be compatible or will be removed.
```
